File: src/main/java/com/github/daggerok/RestResource.java
package com.github.daggerok;

import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

import static javax.ws.rs.core.MediaType.APPLICATION_JSON;

@Path("/v1/hello")
@Produces(APPLICATION_JSON)
public class RestResource {

  @Inject
  GreetingService greetingService;

  @GET
  @Path("")
  public CompletionStage<String> index() {
    return CompletableFuture.supplyAsync(() -> "hello");
  }

  @GET
  @Path("{name}")
  public CompletionStage<String> hello(@PathParam("name") String name) {
    return null == name || name.isEmpty() ? index()
        : CompletableFuture.supplyAsync(() -> greetingService.greeting(name));
  }
}

File: src/main/java/com/github/daggerok/hello/RestResource.scala
package com.github.daggerok.hello

import javax.inject.Inject
import javax.ws.rs._
import java.util.concurrent.CompletionStage

import javax.ws.rs.core.MediaType.APPLICATION_JSON

@Path("/v1/hello")
@Produces(Array(APPLICATION_JSON))
class RestResource {

  @Inject //@scala.beans.BeanProperty
  var greetingService: GreetingService = _

  @GET @Path("")
  def index: CompletionStage[String] =
    greetingService.greetingNoArg

  @GET @Path("{name}")
  def hello(@PathParam("name") name: String): CompletionStage[String] =
    greetingService.greeting(name)
      .exceptionally(e => e.getLocalizedMessage)
}
